### PR TITLE
Support custom error messages/format strings

### PR DESCRIPTION
**Pull request by alazier**
@cmelchior @tgoyne

### CODE REVIEW

File: Realm/ObjectStore/object_store_exceptions.hpp [hunk 1]
```diff
@@ -23,43 +23,77 @@
 #include <map>
 #include <string>
 
+#define INFO_KEY(key) InfoKey InfoKey##key = "InfoKey" #key;
```
Comment (segiddins) [inline Realm/ObjectStore/object_store_exceptions.hpp]: this should definitely be `undef`ed after it's used

**Comment by cmelchior** [comment]
I think this looks like an okay approach. No matter how we do it, we will end up with a lot of strings that has been managed somehow, and this way the strings are very readable.

**Comment by jpsim** [comment]
I actually like this overall for consolidating exception messages, keeping things a bit more typed, and moving towards more shared code in the bindings. Well done, @alazier :+1:.

**Comment by jpsim** [comment]
I added an object store style exception for property optionality change. @alazier can you please confirm this was done the right way?

**Comment by jpsim** [comment]
Changed our minds about this design. Closing.

**Pull request closed without merging**